internal/datatransfer/metrics.go

```diff
@@ -0,0 +1,61 @@
+package datatransfer
+
+import (
+	"log"
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const (
+	DataSyncTransferMetricsAddress           = "localhost:9101"
+	DataSyncTransferMetricsPath              = "/metrics"
+	MetricsEndpoint                          = "http://" + DataSyncTransferMetricsAddress + DataSyncTransferMetricsPath
+	egress                         direction = "egress"
+	ingress                        direction = "ingress"
+)
+
+type direction string
+
+var (
+	registry = prometheus.NewRegistry()
+	factory  = promauto.With(registry)
+
+	FilesTransferredCounter = factory.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "flotta_agent_datasync_files_transferred_counter",
+			Help: "The total of files uploaded for a workload",
+		}, []string{"workload_name", "direction"})
+	BytesTransferredCounter = factory.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "flotta_agent_datasync_bytes_transferred_counter",
+			Help: "The total number of bytes uploaded for a workload",
+		}, []string{"workload_name", "direction"})
+	TimeTransferredCounter = factory.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "flotta_agent_datasync_time_transferred_counter",
+			Help: "The total amount of time spent transferring bytes for a workload",
+		}, []string{"workload_name", "direction"})
+	DeletedRemoteFilesTransferredCounter = factory.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "flotta_agent_datasync_deleted_files_transferred_counter",
+			Help: "The total number of files deleted from the target destination",
+		}, []string{"workload_name", "direction"})
+)
+
+func StartMetrics() {
+	http.Handle(DataSyncTransferMetricsPath, promhttp.Handler())
```

Comment:
Why is HTTP server needed here?

Reply:
Yes we could, but should we deviate from the standard pattern? what is the benefit here? I'm fine doing it if there is a significant gain, otherwise it just adds an exception to what one would expect.

Reply:
I think it would be good to quantify the amount of resources consumed by this handler, if it's idle and it does not consume anything, we should let it here, otherwise we would need to estimate the amount of resources consumed to create a metric and the frequency of the requests to get metrics.
In the device-worker, as we are in a resources' constrained environment, we need to be careful with them, I guess that why @jakub-dzon point that out

Reply:
I'll check if we can call the Prometheus logic to serialize the metrics. Otherwise I'll quantify the extra memory of running the HTTP server inside a go program and report again. I don't think the HTTP server will consume much though but numbers talk for themselves :smile: 

If resource consumption are our concern, maybe we should look at splitting this part from the agent and running it as a container so we can limit the memory and CPU consumption and kill the process if it starts hogging too much memory. Knowing Prometheus a bit, it is the most common explanation why processes start consuming more memory over time, due to cardinality.

At the same time, if we're not doing it already, we should consider running node exporter as a containerized workload and align the metrics generated to those that are to be collected/white listed. This should reduce the memory used by node exporter compared to the default deployment, and as a bonus we also control how much memory it uses.
https://github.com/prometheus/node_exporter#collectors

Thoughts?

Reply:
I do not understand the need to serialize metrics. Most applications create and collect their own metrics. Then these are exposed (exported) via a Prometheus client library. The library's internal implementation is an HTTP server that waits for Prometheus to scrape it. However, we do not have a Prometheus scraping us. We write to TSDB directly. So all you need is to periodically call our metrics 'write'/'add' method. If you generalize it a bit, then we could use it for all our features. Suppose you have a global struct with all metrics. Data upload, logs collection etc. Now suppose you have a timer that triggers every 60 seconds and writes the values from the struct to our TSDB.

Reply:
@ydayagi Is your suggestion to use https://github.com/project-flotta/flotta-device-worker/blob/be75e81fffbda7b0dcbc9c584535fdd0be214677/internal/metrics/daemon.go#L94 
in order to persist the collected sync-usage directly to TSDB and then use the existing remote write to push the metrics.
Any special consideration for this? does the write to TSDB need to be synced or verify it is initialized?

Reply:
use Store, indeed. it is already synced by Prometheus's TSDB. And we already have concurrent calls to Store. Each target is a separate goroutine that calls Stor.

Reply:
Removed httpd server and resolved to directly obtaining the metric data from the prometheus register. @ydayagi @jakub-dzon please take a look.